File: production/client/src/main/scala/kpn/client/filter/Filters.scala
// TODO migrate to Angular
package kpn.client.filter

class Filters[T](filters: Filter[T]*) {

  def passes(element: T): Boolean = {
    passesAll(element, filters)
  }

  def filterExcept(elements: Seq[T], filter: Filter[T]): Seq[T] = {
    filtered(elements, filters.filterNot(_.name == filter.name))
  }

  private def filtered(elements: Seq[T], filterCollection: Seq[Filter[T]]): Seq[T] = {
    elements.filter(element => passesAll(element, filterCollection))
  }

  private def passesAll(element: T, filterCollection: Seq[Filter[T]]): Boolean = {
    !filterCollection.exists(f => !f.passes(element))
  }
}

File: client/libs/shared/src/lib/kpn/filter/filters.ts
import { Filter } from './filter';
import { List } from 'immutable';

export class Filters<T> {
  readonly filters: List<Filter<T>>;

  constructor(...filters: Filter<T>[]) {
    this.filters = List(filters);
  }

  passes(element: T): boolean {
    return this.passesAll(element, this.filters);
  }

  filterExcept(elements: T[], filter: Filter<T>): T[] {
    return this.filtered(
      elements,
      this.filters.filterNot((f) => f.name === filter.name)
    );
  }

  private filtered(elements: T[], filterCollection: List<Filter<T>>): T[] {
    return elements.filter((element) =>
      this.passesAll(element, filterCollection)
    );
  }

  private passesAll(element: T, filterCollection: List<Filter<T>>): boolean {
    return filterCollection.findIndex((f) => !f.passes(element)) < 0;
  }
}
